packages/backend/src/tokens/authStatus.ts

```diff
@@ -70,22 +110,43 @@ export const AuthErrorReason = {
   SessionTokenIatInTheFuture: 'session-token-iat-in-the-future',
   SessionTokenWithoutClientUAT: 'session-token-but-no-client-uat',
   ActiveOrganizationMismatch: 'active-organization-mismatch',
+  TokenTypeMismatch: 'token-type-mismatch',
   UnexpectedError: 'unexpected-error',
 } as const;
 
 export type AuthErrorReason = (typeof AuthErrorReason)[keyof typeof AuthErrorReason];
 
 export type AuthReason = AuthErrorReason | TokenVerificationErrorReason;
 
-export type RequestState = SignedInState | SignedOutState | HandshakeState;
+export type RequestState<T extends TokenType = SessionTokenType> =
+  | AuthenticatedState<T>
+  | UnauthenticatedState<T>
+  | (T extends SessionTokenType ? HandshakeState : never);
+
+type BaseSignedInParams = {
+  authenticateContext: AuthenticateContext;
+  headers?: Headers;
+  token: string;
+  tokenType: TokenType;
+};
+
+type SignedInParams =
+  | (BaseSignedInParams & { tokenType: SessionTokenType; sessionClaims: JwtPayload })
+  | (BaseSignedInParams & { tokenType: MachineTokenType; machineData: MachineAuthType });
+
+export function signedIn<T extends TokenType>(params: SignedInParams & { tokenType: T }): AuthenticatedState<T> {
+  const { authenticateContext, headers = new Headers(), token } = params;
+
+  const toAuth = (() => {
+    if (params.tokenType === TokenType.SessionToken) {
+      const { sessionClaims } = params as { sessionClaims: JwtPayload };
+      return signedInAuthObject(authenticateContext, token, sessionClaims);
+    }
+
+    const { machineData } = params as { machineData: MachineAuthType };
+    return authenticatedMachineObject(params.tokenType, token, machineData, authenticateContext);
+  }) as ToAuth<T, true>;
```

Comment:
Both `signedIn` and signedOut` functions has been updated to accept a `tokenType`, which will then be used to return the correct auth object based on it. If the `tokenType` is not a session_token, it will accept a `machineData` which is the response from calling bapi-proxy `/verify` endpoint.